### PR TITLE
Reduce bound input function in Interactive calls to one

### LINKED ISSUE #763: Bound input function called many times on widget change
I'm preparing the blog post for hvPlot release and running the following example I found out that the load function was executed many times (more than a dozen) on the *Species* widget change.

```python
import panel as pn
import hvplot.pandas  # noqa
from bokeh.sampledata import penguins

pn.extension()

df = penguins.data

w_species = pn.widgets.Select(name='Species', options=list(df['species'].unique()))

w_sex = pn.widgets.MultiSelect(name='Sex', value=['MALE'], options=['MALE', 'FEMALE'])
w_body_mass = pn.widgets.FloatSlider(name='Min body mass', start=df['body_mass_g'].min(), end=df['body_mass_g'].max(), step=50)
w_y_dim = pn.widgets.RadioButtonGroup(options=['bill_depth_mm', 'flipper_length_mm']) 


def load_data(species):
    """Simluate loading data from e.g a database or from a web API."""
    print('species')
    data = df.loc[df['species'] == species]
    # time.sleep(2)
    # print(species)
    return data

dfi = hvplot.bind(load_data, w_species).interactive()

interactive = (
    dfi
    .loc[(dfi['sex'].isin(w_sex)) & (dfi['body_mass_g'] > w_body_mass)]
    .hvplot(kind='scatter', x='bill_length_mm', y=w_y_dim, by='island')
)

pn.Column(
    interactive.widgets(),
    interactive.panel(),
).servable()
```

image

### PR DESCRIPTION
**Pull request by Hoxbro**
Fixes #763

Most extra function calls were done inside `__init__`. I have added the `Watcher` to the `watchers` list if it is not already there, and if it is on the list, I `unwatch` it. I have a suspicion this is not the best practice and could be done more elegantly.  

The change in `__new__` reduces the bound function call from two to one (after the change in `__init__`).

### CODE REVIEW

**Comment by Hoxbro** [comment]
I get the following warning when running the script with `python file.py`. The warning is only there when `.hvplot` is added to interactive.

``` python
import panel as pn
import hvplot.pandas  # noqa
from bokeh.sampledata import penguins

pn.extension()

df = penguins.data
w_species = pn.widgets.Select(name='Species', options=list(df['species'].unique()))
w_sex = pn.widgets.MultiSelect(name='Sex', value=['MALE'], options=['MALE', 'FEMALE'])


def load_data(species):
    """Simluate loading data from e.g a database or from a web API."""
    print('species')
    data = df.loc[df['species'] == species]
    return data

df2 = hvplot.bind(load_data, w_species)
dfi = df2.interactive()

interactive = (
    dfi
    .loc[(dfi['sex'].isin(w_sex))]
    .hvplot(kind='scatter')  # No warning if this is commented out
)
```

``` python-traceback
Exception ignored in: functools.partial(<function cleanup_custom_options at 0x7fd3213205e0>, 2)
Traceback (most recent call last):
  File "/home/shh/Development/holoviz/holoviews/holoviews/core/options.py", line 81, in cleanup_custom_options
Exception: Cleanup of custom options tree with id '2' failed with the following exception: 'NoneType' object has no attribute '_options_context', an unreferenced orphan tree may persist in memory.
Exception ignored in: functools.partial(<function cleanup_custom_options at 0x7fd3213205e0>, 2)
Traceback (most recent call last):
  File "/home/shh/Development/holoviz/holoviews/holoviews/core/options.py", line 81, in cleanup_custom_options
Exception: Cleanup of custom options tree with id '2' failed with the following exception: 'NoneType' object has no attribute '_options_context', an unreferenced orphan tree may persist in memory.
Exception ignored in: functools.partial(<function cleanup_custom_options at 0x7fd3213205e0>, 2)
Traceback (most recent call last):
  File "/home/shh/Development/holoviz/holoviews/holoviews/core/options.py", line 81, in cleanup_custom_options
Exception: Cleanup of custom options tree with id '2' failed with the following exception: 'NoneType' object has no attribute '_options_context', an unreferenced orphan tree may persist in memory.
Exception ignored in: functools.partial(<function cleanup_custom_options at 0x7fd3213205e0>, 2)
Traceback (most recent call last):
  File "/home/shh/Development/holoviz/holoviews/holoviews/core/options.py", line 81, in cleanup_custom_options
Exception: Cleanup of custom options tree with id '2' failed with the following exception: 'NoneType' object has no attribute '_options_context', an unreferenced orphan tree may persist in memory.
```

**Comment by maximlt** [comment]
@Hoxbro this is the code snippet I've been running in a notebook and you should see that the output, here the shape of the dataframe, doesn't update with the current changes of this PR while it should:

```python
import panel as pn
import hvplot.pandas
from bokeh.sampledata import airports

df = airports.data

def foo(max_elev):
    df = airports.data
    return df.loc[df.elevation < max_elev]

w = pn.widgets.IntSlider(value=2000, start=1000, end=3000)
dfi = hvplot.bind(foo, w).interactive()
dfi.shape
```

example_hvplot_interactive

**Comment by maximlt** [comment]
@philippjfr could you have a look at this PR? What happens is that many watchers are registered to call the input function when the input widget(s) change. This happens at this location in the code, because `Interactive` is cloned quite a few times when the pipeline is built: 

https://github.com/holoviz/hvplot/blob/ce0f4e135e0bb0c940a5092edb6d9436377b9c70/hvplot/interactive.py#L83-L85

In general it would be great if you could add a few comments in the `Interactive` code, for us to better understand its methods but also its overall logic.

**Comment by philippjfr** [comment]
Hmm, yeah this one is tricky. I wonder if an easier fix is simply not to inherit the `_fn` when chaining the pipeline. Without diving in I'm not sure how straightforward that is however.

File: hvplot/interactive.py [hunk 1]
```diff
@@ -169,7 +169,16 @@ def _clone(self, transform=None, plot=None, loc=None, center=None,
             kwargs = dict(self._kwargs, inherit_kwargs=self._inherit_kwargs, method=self._method, **kwargs)
         else:
             kwargs = dict(self._inherit_kwargs, **dict(self._kwargs, **kwargs))
-        return type(self)(self._obj, fn=self._fn, transform=transform, plot=plot, depth=depth,
+
+        # This is a hack to avoid calling the original bound function multiple times
+        # when cloning. self._obj is updated by the orignal function with depth=0
+        if self._fn:
+            f = pn.panel(pn.bind(lambda *args, **kwargs: self._obj), lazy=True)
+            f.object._dinfo = self._fn.object._dinfo
```
Comment (philippjfr) [inline hvplot/interactive.py:177]: Very clever but also additional layers of indirection. Let's go with https://github.com/holoviz/hvplot/pull/818

**Comment by maximlt** [comment]
Superseded by https://github.com/holoviz/hvplot/pull/818

File: hvplot/interactive.py (hunk 1 above)
Comment (Hoxbro) [reply hvplot/interactive.py:177]: I completely agree that #818 is the better choice.

**Pull request closed without merging**